File: riveredge-backend/src/apps/kuaizhizao/schemas/customer_pool.py
"""
客户池 API Schema。
"""

from __future__ import annotations

from datetime import datetime
from typing import List, Optional

from pydantic import BaseModel, ConfigDict, Field


class CustomerPoolCollaboratorItem(BaseModel):
    user_id: int
    user_name: str

    model_config = ConfigDict(from_attributes=True)


class CustomerPoolItem(BaseModel):
    id: int
    uuid: str
    code: str
    name: str
    short_name: Optional[str] = None
    contact_person: Optional[str] = None
    phone: Optional[str] = None
    salesman_id: Optional[int] = None
    salesman_name: Optional[str] = None
    pool_status: str
    assigned_at: Optional[datetime] = None
    last_follow_up_at: Optional[datetime] = None
    recycle_at: Optional[datetime] = None
    created_by_name: Optional[str] = None
    updated_by_name: Optional[str] = None
    created_at: datetime
    updated_at: datetime
    collaborators: List[CustomerPoolCollaboratorItem] = Field(default_factory=list)

    model_config = ConfigDict(from_attributes=True)


class CustomerPoolListEnvelope(BaseModel):
    items: List[CustomerPoolItem]
    total: int


class CustomerPoolAssignBody(BaseModel):
    salesman_id: int = Field(..., ge=1, description="目标业务员ID")
    reason: Optional[str] = Field(None, max_length=200)


class CustomerPoolActionBody(BaseModel):
    reason: Optional[str] = Field(None, max_length=200)


class CustomerPoolRuleResponse(BaseModel):
    recycle_enabled: bool = True
    recycle_after_days: int = 15
    max_owned_customers: int = 0
    allow_claim_others: bool = False
    updated_at: Optional[datetime] = None
    updated_by: Optional[int] = None

    model_config = ConfigDict(from_attributes=True)


class CustomerPoolRuleUpdateBody(BaseModel):
    recycle_enabled: Optional[bool] = None
    recycle_after_days: Optional[int] = Field(None, ge=1, le=365)
    max_owned_customers: Optional[int] = Field(None, ge=0, le=100000)
    allow_claim_others: Optional[bool] = None


class CustomerPoolCollaboratorsUpdateBody(BaseModel):
    user_ids: List[int] = Field(default_factory=list, description="协作人用户ID列表")


class CustomerPoolLogItem(BaseModel):
    action: str
    from_salesman_id: Optional[int] = None
    from_salesman_name: Optional[str] = None
    to_salesman_id: Optional[int] = None
    to_salesman_name: Optional[str] = None
    operator_user_id: int
    operator_name: Optional[str] = None
    reason: Optional[str] = None
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class CustomerPoolLogListEnvelope(BaseModel):
    items: List[CustomerPoolLogItem]
    total: int

File: riveredge-frontend/src/apps/kuaizhizao/services/customer-pool.ts
import { apiRequest } from '../../../services/api';

export interface CustomerPoolCollaboratorItem {
  user_id: number;
  user_name: string;
}

export interface CustomerPoolItem {
  id: number;
  uuid: string;
  code: string;
  name: string;
  short_name?: string | null;
  contact_person?: string | null;
  phone?: string | null;
  salesman_id?: number | null;
  salesman_name?: string | null;
  pool_status: 'pool' | 'owned';
  assigned_at?: string | null;
  last_follow_up_at?: string | null;
  recycle_at?: string | null;
  created_by_name?: string | null;
  updated_by_name?: string | null;
  created_at: string;
  updated_at: string;
  collaborators?: CustomerPoolCollaboratorItem[];
}

export interface CustomerPoolListResult {
  items: CustomerPoolItem[];
  total: number;
}

export interface CustomerPoolRule {
  recycle_enabled: boolean;
  recycle_after_days: number;
  max_owned_customers: number;
  allow_claim_others: boolean;
  updated_at?: string | null;
  updated_by?: number | null;
}

export interface CustomerPoolLogItem {
  action: string;
  from_salesman_id?: number | null;
  from_salesman_name?: string | null;
  to_salesman_id?: number | null;
  to_salesman_name?: string | null;
  operator_user_id: number;
  operator_name?: string | null;
  reason?: string | null;
  created_at: string;
}

export interface CustomerPoolLogListResult {
  items: CustomerPoolLogItem[];
  total: number;
}

export interface CustomerPoolListParams {
  scope?: 'pool' | 'mine' | 'all';
  skip?: number;
  limit?: number;
  keyword?: string;
  code?: string;
  name?: string;
  contact_person?: string;
  phone?: string;
  salesmanId?: number;
  poolStatus?: 'pool' | 'owned';
  last_follow_up_from?: string;
  last_follow_up_to?: string;
  recycle_from?: string;
  recycle_to?: string;
  assigned_from?: string;
  assigned_to?: string;
  created_start_date?: string;
  created_end_date?: string;
  updated_start_date?: string;
  updated_end_date?: string;
  order_by?: string;
}

export const customerPoolApi = {
  list: async (params?: CustomerPoolListParams): Promise<CustomerPoolListResult> =>
    apiRequest('/apps/kuaizhizao/customer-pool', { method: 'GET', params }),

  claim: async (customerId: number, reason?: string): Promise<CustomerPoolItem> =>
    apiRequest(`/apps/kuaizhizao/customer-pool/${customerId}/claim`, {
      method: 'POST',
      data: { reason },
    }),

  assign: async (customerId: number, salesmanId: number, reason?: string): Promise<CustomerPoolItem> =>
    apiRequest(`/apps/kuaizhizao/customer-pool/${customerId}/assign`, {
      method: 'POST',
      data: { salesman_id: salesmanId, reason },
    }),

  release: async (customerId: number, reason?: string): Promise<CustomerPoolItem> =>
    apiRequest(`/apps/kuaizhizao/customer-pool/${customerId}/release`, {
      method: 'POST',
      data: { reason },
    }),

  recycle: async (customerId: number, reason?: string): Promise<CustomerPoolItem> =>
    apiRequest(`/apps/kuaizhizao/customer-pool/${customerId}/recycle`, {
      method: 'POST',
      data: { reason },
    }),

  getRules: async (): Promise<CustomerPoolRule> =>
    apiRequest('/apps/kuaizhizao/customer-pool/rules', { method: 'GET' }),

  updateRules: async (data: Partial<CustomerPoolRule>): Promise<CustomerPoolRule> =>
    apiRequest('/apps/kuaizhizao/customer-pool/rules', { method: 'PUT', data }),

  getCollaborators: async (customerId: number): Promise<CustomerPoolCollaboratorItem[]> =>
    apiRequest(`/apps/kuaizhizao/customer-pool/${customerId}/collaborators`, { method: 'GET' }),

  setCollaborators: async (customerId: number, userIds: number[]): Promise<CustomerPoolCollaboratorItem[]> =>
    apiRequest(`/apps/kuaizhizao/customer-pool/${customerId}/collaborators`, {
      method: 'PUT',
      data: { user_ids: userIds },
    }),

  getLogs: async (customerId: number): Promise<CustomerPoolLogListResult> =>
    apiRequest(`/apps/kuaizhizao/customer-pool/${customerId}/logs`, { method: 'GET' }),
};

